### PR TITLE
fix: properly close osm & emit close event

### LINKED ISSUE #82: Potential race condition with swarm
If  [`sync.listen()`](https://github.com/digidem/mapeo-core/blob/master/sync.js#L225-L232) is called while the discovery swarm is in the process of being destroyed, then it will not create new swarm:

```js
  listen (cb) {
    if (!cb) cb = () => {}
    if (this.swarm || this._destroyingSwarm) {
      return process.nextTick(cb)
    }
    this.swarm = this._swarm()
    process.nextTick(cb)
  }
```

Also, if [`sync.close()`](https://github.com/digidem/mapeo-core/blob/master/sync.js#L248-L257) is called a second time while the sync is already closing, the callback could be called before the sync swarm is destroyed (this seems minor and unlikely to create many problems, but could do in combination with the bug above):

```js
  close (cb) {
    if (!cb) cb = () => {}
    if (!this.swarm || this._destroyingSwarm) return process.nextTick(cb)
    this._destroyingSwarm = true
    this.swarm.destroy(() => {
      this.swarm = null
      this._destroyingSwarm = false
      cb()
    })
  }
```

This can be an issue because on Android the server needs to close when the app goes into the background (otherwise it drains the battery, and Android can force close the app). When Android requests user permission (e.g. for Camera use) or opens the share dialog, then this also puts the app into the background, which means that `close()` and `listen()` can be called quickly in succession.

### PR DESCRIPTION
**Pull request by okdistribute**
This back ports fixes to close osm from v9.

Also fixes #82

### CODE REVIEW

File: sync.js [hunk 1]
```diff
@@ -217,12 +217,18 @@ class Sync extends events.EventEmitter {
...
     }
-    this.swarm = this._swarm()
-    this.swarm.listen(0, cb)
+
+    var _listen = () => {
+      this.swarm = this._swarm()
+      this.swarm.listen(0, cb)
+    }
+
+    if (this._destroyingSwarm) this.on('close', _listen)
```
Comment (noffle) [inline sync.js:230]:
```suggestion
    if (this._destroyingSwarm) this.once('close', _listen)
```

File: sync.js [hunk 2]
```diff
@@ -217,12 +217,18 @@ class Sync extends events.EventEmitter {
 
   listen (cb) {
     if (!cb) cb = () => {}
-    if (this.swarm || this._destroyingSwarm) {
-      console.error('Swarm already exists or is currently destroying itself..')
+    if (this.swarm && !this._destroyingSwarm) {
+      console.error('Swarm already exists.')
```
Comment (noffle) [inline sync.js:221]:
```suggestion
```

**Comment by okdistribute** [comment]
The failures are expected, right?

**Comment by noffle** [comment]
@okdistribute no, these failures look new! the last commit to `v8` [passed](https://travis-ci.org/github/digidem/mapeo-core/builds/680306164)

**Comment by noffle** [comment]
I looked into this and I saw two things:

1. the 2nd lifecycle test calls "close" twice, which the code isn't set up to handle properly
2. a *lot* of confounding issues are coming, I think, from the various race conditions in the sync tests

I suggest folding this change into #87, where I've applied some more sync test fixes to get everything passing again. You ok if we close this?

**Pull request closed without merging**